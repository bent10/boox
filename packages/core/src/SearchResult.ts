import type { Encoder } from './Encoder.js'
import type {
  Dataset,
  Document,
  QueryContext,
  SearchResultOptions
} from './types.js'

/**
 * Represents the result of a full-text search operation, providing information
 * about the search query's impact on the document.
 */
export class SearchResult<T extends object = Dataset> {
  /**
   * The search query.
   */
  query: string

  /**
   * The reference identifier for the document.
   */
  id: Document<T>['id']

  /**
   * The document elaborates on the fields within the dataset.
   */
  attributes: Document<T>['attributes']

  /**
   * The document magnitude.
   */
  magnitude: Document<T>['magnitude']

  /**
   * The score of the search result.
   */
  score: number

  /**
   * The HTML tag markers for keyword highlighting.
   */
  highlightTag: [string, string]

  vector: {
    dotProducts: number[]
    queryMagnitudeSquare: number[]
  } | null

  /**
   * The model encoder.
   */
  #encoder: Encoder

  /**
   * Creates a new SearchResult instance.
   *
   * @param options - The configuration options for the search result.
   */
  constructor(options: SearchResultOptions<T>) {
    const {
      encoder,
      query,
      document,
      highlightTag = ['<mark class="search-highlight">', '</mark>']
    } = options

    this.#encoder = encoder

    this.query = query
    this.id = document.id
    this.attributes = document.attributes
    this.magnitude = document.magnitude
    this.score = 0
    this.highlightTag = highlightTag
    this.vector = null
  }

  /**
   * Retrieves the context of the specified document attribute.
   *
   * @param key - The key of the document attribute to retrieve context for.
   * @param contextLength - The length of the returned context.
   * @returns The context of search results for a document, including
   *   keywords and their context.
   */
  context(key: keyof T, contextLength?: number): QueryContext {
    const text = <string>this.attributes[key] || ''

    if (!contextLength) {
      contextLength = text.length
    }

    const keywords = new Set<string>()
    let chunk = ''
    let index = 0

    while (index < text.length) {
      chunk = text.slice(index, index + contextLength)
      const indices = this.#findIndices(chunk)

      indices.map(({ keyword }) => keywords.add(keyword))

      if (keywords.size) break

      index += contextLength
    }

    return {
      keywords,
      text: this.#highlightSearchTerms(chunk, keywords)
    }
  }

  /**
   * Retrieves keyword in context (KWIC) results for the specified document
   * attribute.
   *
   * @param key - The key of the document attribute to retrieve KWIC results
   *   for.
   * @param wordsAround The number of words to retrieve before and
   *   after the keyword.
   * @param isSorted - Whether to sort the results by keyword size.
   * @returns An array of KWIC results for a document, including keywords
   *   and their context.
   */
  kwic(key: keyof T, wordsAround = 7, isSorted = false) {
    const results: QueryContext[] = []
    const text = <string>this.attributes[key] || ''

    if (!text) {
      return results
    }

    const normalizedText = text
      .split(/\r\n|\r|\n/)
      .filter(line => line.trim() !== '')
      .join(' ')

    this.#findIndices(normalizedText).forEach(({ index }) => {
      const context = this.#getWordsAroundIndex(
        normalizedText,
        index,
        wordsAround
      )

      if (context) {
        const keywords = new Set(this.#extractKeywords(context))

        results.push({
          keywords,
          text: this.#highlightSearchTerms(context, keywords)
        })
      }
    })

    return isSorted
      ? results.sort((a, b) => b.keywords.size - a.keywords.size)
      : results
  }

  /**
   * Highlights search terms within the text.
   *
   * @param text - The text to highlight search terms in.
   * @returns The text with search terms highlighted.
   */
  #highlightSearchTerms(
    text: string,
    fallbackKeywords: QueryContext['keywords']
  ) {
    const escapedQuery = this.#escapeRegExp(this.query)
    const pattern =
      new RegExp(escapedQuery, 'i').test(text) || !fallbackKeywords.size
        ? escapedQuery
        : Array.from(fallbackKeywords)
            .map(k => this.#escapeRegExp(k))
            .join('|')

    const regex = new RegExp(`(${pattern})`, 'gi')
    const [openTag, closeTag] = this.highlightTag

    return text
      .replace(/^\W+/, '')
      .replace(regex, `${openTag}$1${closeTag}`)
      .trimEnd()
  }

  /**
   * Retrieves the specified number of words before and after the given index in
   * a text.
   *
   * @param text The large text.
   * @param index The index around which to retrieve words.
   * @param wordsAround The number of words to retrieve before and
   *   after the index.
   * @returns A string containing the specified number of words before and after the given index, separated by spaces.
   */
  #getWordsAroundIndex(text: string, index: number, wordsAround: number) {
    const wordBoundaryRegex =
      /(\b(?:\w|(?:&(?:[a-z0-9]+|#[0-9]{1,6}|#x[0-9a-f]{1,6});))+\b)/gi
    const words = text.match(wordBoundaryRegex)

    if (words) {
      const wordContainingIndex = words.find(
        word =>
          text.indexOf(word) <= index &&
          text.indexOf(word) + word.length >= index
      )

      if (wordContainingIndex) {
        const wordArrayIndex = words.indexOf(wordContainingIndex)

        // Calculate the start and end indices for the specified number of words before and after
        const start = Math.max(0, wordArrayIndex - wordsAround)
        const end = Math.min(words.length, wordArrayIndex + wordsAround + 1)

        return words.slice(start, end).join(' ')
      }
    }
  }

  /**
   * Finds the indices of query parts within the given text.
   *
   * @param text The text to search within.
   * @returns An array of indices where query parts are found within the text.
   */
  #findIndices(text: string) {
    const indices: Array<{ keyword: string; index: number }> = []
    const keywords = this.#extractKeywords(text)

    const textLowerCase = text.toLowerCase()
    let pointer = 0

    for (const keyword of keywords) {
      const index = textLowerCase.slice(pointer).indexOf(keyword.toLowerCase())

      if (index !== -1) {
        indices.push({ keyword, index })
        pointer = index + keyword.length
      }
    }

    return indices
  }

  /**
   * Extracts keywords from the given text that are similar to the search query.
   *
   * @param text The text to extract keywords from.
   * @returns An array of extracted keywords.
   */
  #extractKeywords(text: string): string[] {
    const qEncodings = this.#encoder.encode(this.query)
    const normalizedText = this.#encoder.normalize(text)

    return this.#encoder.tokenize(normalizedText).filter(keyword => {
      const kEncodings = this.#encoder.encode(keyword)
      return this.#areEncodingsSimilar(qEncodings, kEncodings)
    })
  }

  /**
   * Compares two encodings to check if they have at least one matching
   * element.
   *
   * @param encodings1 The first encodings to compare.
   * @param encodings2 The second encodings to compare.
   * @returns True if the encodings are similar, false otherwise.
   */
  #areEncodingsSimilar(encodings1: string[], encodings2: string[]) {
    const set = new Set(encodings1)
    for (const item of encodings2) {
      if (set.has(item)) {
        return true
      }
    }
    return false
  }

  /**
   * Escapes special characters in a string for use in a regular expression.
   *
   * @param string The string to escape.
   * @returns The escaped string.
   */
  #escapeRegExp(pattern: string) {
    return pattern.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&').replace(/-/g, '\\x2d')
  }
}
