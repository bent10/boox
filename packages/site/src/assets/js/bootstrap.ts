import { Modal, Tooltip } from 'bootstrap'

// Constants
const TOOLTIP_SELECTOR = '[data-bs-toggle="tooltip"]'

/**
 * Initializes Bootstrap UI components.
 */
export function initBootstrapUI() {
  initializeTooltips()
}

/**
 * Initializes tooltip components.
 */
function initializeTooltips() {
  document.querySelectorAll(TOOLTIP_SELECTOR).forEach(el => {
    Tooltip.getOrCreateInstance(el)
  })
}

/**
 * Handles keyboard shortcuts for showing a modal.
 *
 * @param selector - The selector for the modal element.
 * @param key - The key to trigger the modal (default: '/').
 */
export function handleModalKeydown(selector: string, key = '/') {
  document.addEventListener('keydown', e => {
    if (e.key === key) {
      const modalElement = document.querySelector<HTMLElement>(selector)
      if (modalElement) {
        Modal.getOrCreateInstance(modalElement).show(modalElement)
      }
    }
  })
}
