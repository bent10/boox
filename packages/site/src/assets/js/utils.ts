import type { State } from 'boox'

/**
 * Fetches data from the specified URL asynchronously and caches it until
 * the user closes the browser.
 *
 * @param url The URL from which to fetch the data.
 * @returns A Promise that resolves to the fetched data.
 * @throws
 *   Error if the network response was not OK or an error occurred during
 *   fetching.
 */
export async function fetchData<T extends object = object>(url: string) {
  try {
    const response = await fetch(url, {
      method: 'GET',
      mode: 'cors',
      headers: {
        'Content-Encoding': 'gzip',
        'Content-Type': 'application/json'
      },
      cache: 'default'
    })

    if (!response.ok) {
      throw new Error('Network response was not ok')
    }

    const blob = await response.blob()
    const ds = new DecompressionStream('gzip')
    const decompressedStream = blob.stream().pipeThrough(ds)

    return (await new Response(decompressedStream).json()) as State<T>
  } catch (error) {
    console.error(error)
    throw error
  }
}

/**
 * Creates a debounced version of a function that delays invoking the original
 * function until after `delay` milliseconds have elapsed since the last time
 * the debounced function was invoked.
 *
 * @param func - The function to debounce.
 * @param delay - The number of milliseconds to delay.
 * @returns A debounced function.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function debounce<F extends (...args: any[]) => any>(
  func: F,
  delay: number
): (...args: Parameters<F>) => Promise<ReturnType<F>> {
  let timeoutId: ReturnType<typeof setTimeout>

  return function debouncedFunction(
    this: ThisParameterType<F>,
    ...args: Parameters<F>
  ): Promise<ReturnType<F>> {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this

    clearTimeout(timeoutId)

    return new Promise<ReturnType<F>>(resolve => {
      timeoutId = setTimeout(async () => {
        const result = await func.apply(context, args)
        resolve(result)
      }, delay)
    })
  }
}
