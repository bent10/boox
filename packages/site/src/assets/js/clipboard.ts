import { Tooltip } from 'bootstrap'
import ClipboardJS from 'clipboard'

const ICON_SUCCESS =
  '<svg xmlns="http://www.w3.org/2000/svg" width="1rem" height="1rem" viewBox="0 0 24 24"><g><path d="M22.641,4.232c-.427-.354-1.056-.296-1.409,.128L11.933,15.519l-.433-.433-1.286,1.543,1.078,1.078c.188,.188,.442,.293,.707,.293,.015,0,.03,0,.045,0,.281-.013,.543-.143,.724-.359L22.769,5.64c.354-.424,.296-1.055-.128-1.408Z" fill="var(--bs-success)"></path><path d="M6,18c-.265,0-.52-.105-.707-.293L1.293,13.707c-.391-.391-.391-1.023,0-1.414s1.023-.391,1.414,0l3.226,3.226L15.231,4.36c.354-.424,.983-.481,1.409-.128,.424,.354,.481,.984,.128,1.408L6.769,17.64c-.181,.216-.442,.346-.724,.359-.015,0-.03,0-.045,0Z" fill="currentColor"></path></g></svg>'

/**
 * Handles clipboard functionality for copying code snippets.
 */
export function handleClipboard() {
  // Initialize ClipboardJS
  const clipboard = new ClipboardJS('[data-js-copy]', {
    target: trigger =>
      trigger.parentElement?.parentElement?.querySelector('code') || trigger
  })

  clipboard.on('success', e => {
    e.clearSelection()

    const trigger = e.trigger as HTMLElement
    const icon = trigger.querySelector('svg')?.outerHTML || 'Copy'

    showSuccessIcon(trigger)
    setTimeout(() => resetTrigger(trigger, icon), 2000)
  })

  clipboard.on('error', e => {
    showFallbackMessage(e.trigger as HTMLElement)
  })
}

/**
 * Displays the success icon in the trigger element and updates the tooltip.
 *
 * @param trigger - The trigger element that initiated the copy action.
 */
function showSuccessIcon(trigger: HTMLElement) {
  trigger.innerHTML = ICON_SUCCESS
  trigger.classList.add('pe-none')

  const tooltip = Tooltip.getInstance(trigger)
  tooltip?.setContent({ '.tooltip-inner': 'Copied!' })
}

/**
 * Resets the trigger element to its original state and hides the tooltip.
 *
 * @param trigger - The trigger element that initiated the copy action.
 */
function resetTrigger(trigger: HTMLElement, icon: string) {
  trigger.innerHTML = icon
  trigger.classList.remove('pe-none')

  const tooltip = Tooltip.getInstance(trigger)
  tooltip?.setContent({ '.tooltip-inner': 'Copy to clipboard' })
  tooltip?.hide()
}

/**
 * Displays a fallback message in the tooltip for unsupported copy actions.
 *
 * @param trigger - The trigger element that initiated the copy action.
 */
function showFallbackMessage(trigger: HTMLElement) {
  const modifierKey = /mac/i.test(navigator.userAgent) ? '\u2318' : 'Ctrl-'
  const fallbackMsg = `Press ${modifierKey}C to copy`
  const tooltip = Tooltip.getInstance(trigger)

  tooltip?.setContent({ '.tooltip-inner': fallbackMsg })
  trigger.addEventListener(
    'hidden.bs.tooltip',
    () => tooltip?.setContent({ '.tooltip-inner': 'Copy to clipboard' }),
    { once: true }
  )
}
