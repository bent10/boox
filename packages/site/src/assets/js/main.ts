import { handleModalKeydown, initBootstrapUI } from './bootstrap.js'
import { handleClipboard } from './clipboard.js'
import { handleSearch, handleClickedSearchResults } from './search.js'
import {
  applyTheme,
  getPreferredTheme,
  handleThemeSwitcher,
  setupColorSchemeListener
} from './theme.js'

initBootstrapUI()
handleModalKeydown('#search-modal', '/')
handleClickedSearchResults('#results', '#search-modal')

handleClipboard()

handleSearch()

applyTheme(getPreferredTheme())
handleThemeSwitcher('[data-bs-theme-switcher]')
setupColorSchemeListener()
