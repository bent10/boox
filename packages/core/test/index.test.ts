/// <reference types="vitest/globals" />

import { readFile } from 'node:fs/promises'
import type { MockInstance } from 'vitest'
import Boox from '../src/index.js'
import { SearchResult } from '../src/SearchResult.js'

type Dataset = {
  id: string
  title: string
  content: string
  url: string
}

describe('Boox', () => {
  let boox: Boox<Dataset>
  let datasets: Dataset[]
  let errorSpy: MockInstance

  beforeEach(() => {
    datasets = [
      {
        id: '1',
        title: 'Document 1',
        content: 'Lorem ipsum dolor sit amet.',
        url: '/foo'
      },
      {
        id: '2',
        title: 'Document 2',
        content: 'Consectetur adipiscing elit.',
        url: '/bar'
      },
      {
        id: '3',
        title: 'Document 3',
        content: 'Sed do eiusmod tempor incididunt.',
        url: '/baz'
      }
    ]
    boox = new Boox<Dataset>({
      id: 'id',
      features: ['title', 'content'],
      attributes: ['url']
    })

    // mock console.error
    errorSpy = vi.spyOn(console, 'error').mockImplementation(() => {})
  })

  afterEach(() => {
    // restore console.error
    errorSpy.mockRestore()
  })

  it('should add a document to the index', async () => {
    await boox.addDocument(datasets[0])

    expect(boox.getDocument('1')).toMatchSnapshot()
  })

  it('should add a document to the index synchronously', () => {
    boox.addDocumentSync(datasets[0])

    expect(boox.getDocument('1')).toMatchSnapshot()
  })

  it('should update an existing document in the index', async () => {
    await boox.addDocument(datasets[0])

    const updatedDataset = {
      ...datasets[0],
      title: 'Updated Document Title',
      content: 'Updated content.'
    }
    await boox.updateDocument(updatedDataset)

    expect(boox.getDocument('1')).toMatchSnapshot()
  })

  it('should update an existing document in the index synchronously', () => {
    boox.addDocumentSync(datasets[0])

    const updatedDataset = {
      ...datasets[0],
      title: 'Updated Document Title',
      content: 'Updated content.'
    }
    boox.updateDocumentSync(updatedDataset)

    expect(boox.getDocument('1')).toMatchSnapshot()
  })

  it('should remove a document from the index', async () => {
    await boox.addDocument(datasets[0])
    const isRemoved = boox.removeDocument('1')

    expect(boox.getDocument('1')).toBeUndefined()
    expect(isRemoved).toBe(true)
  })

  it('should remove a document from the index synchronously', () => {
    boox.addDocumentSync(datasets[0])
    const isRemoved = boox.removeDocument('1')

    expect(boox.getDocument('1')).toBeUndefined()
    expect(isRemoved).toBe(true)
  })

  it('should resolve missing document removal from index', () => {
    const isRemoved = boox.removeDocument('1')

    expect(boox.getDocument('1')).toBeUndefined()
    expect(isRemoved).toBe(false)
  })

  it('should add multiple documents to the index', async () => {
    await boox.addDocuments(datasets)

    expect(boox.getDocument('1')).toMatchSnapshot()
    expect(boox.getDocument('2')).toMatchSnapshot()
    expect(boox.getDocument('3')).toMatchSnapshot()
  })

  it('should add multiple documents to the index synchronously', () => {
    boox.addDocumentsSync(datasets)

    expect(boox.getDocument('1')).toMatchSnapshot()
    expect(boox.getDocument('2')).toMatchSnapshot()
    expect(boox.getDocument('3')).toMatchSnapshot()
  })

  it('should retrieve the current state of the Boox instance', async () => {
    await boox.addDocuments(datasets)

    expect(boox.currentState).toMatchSnapshot()
  })

  it('should set the current state of the Boox instance', async () => {
    const trainedData = JSON.parse(
      await readFile('test/fixtures/dummy-state.json', 'utf8')
    )

    boox.currentState = trainedData

    expect(boox.currentState).toEqual(trainedData)
  })

  it('should search for documents based on a query', async () => {
    await boox.addDocuments(datasets)

    const results = await boox.search('lorem ipsum')

    expect(results.length).toBe(1)
    expect(results[0].id).toBe('1')
  })

  it('should search for documents based on a query synchronously', () => {
    boox.addDocumentsSync(datasets)

    const results = boox.searchSync('lorem ipsum')

    expect(results.length).toBe(1)
    expect(results[0].id).toBe('1')
  })

  it('should handle empty search results', async () => {
    const results = await boox.search('Non-existing document')

    expect(results.length).toBe(0)
  })

  it('should handle empty search results synchronously', () => {
    const results = boox.searchSync('Non-existing document')

    expect(results.length).toBe(0)
  })

  it('should search for documents with custom limit', async () => {
    await boox.addDocuments(datasets)

    const results = await boox.search('document', {
      limit: 2
    })

    expect(results.length).toBe(2)
  })

  it('should search for documents with custom limit synchronously', async () => {
    boox.addDocumentsSync(datasets)

    const results = boox.searchSync('document', {
      limit: 2
    })

    expect(results.length).toBe(2)
  })

  it('should handle search with query expander', async () => {
    await boox.addDocuments(datasets)

    const searchResults = await boox.search('lorem', {
      queryExpander: (query: string) => query.toLowerCase()
    })

    expect(searchResults.length).toBe(1)
    expect(searchResults[0].id).toBe('1')
  })

  it('should handle search with query expander synchronously', () => {
    boox.addDocumentsSync(datasets)

    const searchResults = boox.searchSync('lorem', {
      queryExpander: (query: string) => query.toLowerCase()
    })

    expect(searchResults.length).toBe(1)
  })

  it('should search for documents with vector model', async () => {
    await boox.addDocuments(datasets)

    const results = await boox.search('lorem ipsum', {
      useQueryVector: true
    })

    expect(results.length).toBe(1)
    expect(results[0].id).toBe('1')
  })

  it('should search for documents with vector model synchronously', () => {
    boox.addDocumentsSync(datasets)

    const results = boox.searchSync('lorem ipsum', {
      useQueryVector: true
    })

    expect(results.length).toBe(1)
    expect(results[0].id).toBe('1')
  })

  it('should handle searching with advanced similarity scoring', async () => {
    await boox.addDocuments(datasets)

    const results = await boox.search('document', {
      matchingCoefficient: (
        query: string,
        attributes: Record<string, unknown>
      ) => {
        if (String(attributes.title).toLowerCase().includes(query)) return 0.5
        return 0
      }
    })

    expect(results).toMatchSnapshot()
  })

  it('should handle searching with advanced similarity scoring synchronously', () => {
    boox.addDocumentsSync(datasets)

    const results = boox.searchSync('document', {
      matchingCoefficient: (
        query: string,
        attributes: Record<string, unknown>
      ) => {
        if (String(attributes.title).toLowerCase().includes(query)) return 0.5
        return 0
      }
    })

    expect(results).toMatchSnapshot()
  })

  it('should cache search results for subsequent searches', async () => {
    await boox.addDocuments(datasets)

    const results1 = await boox.search('lorem ipsum')
    const results2 = await boox.search('lorem ipsum')

    expect(results1.length).toBe(1)
    expect(results2.length).toBe(1)
    expect(results1).toEqual(results2)
  })

  it('should cache search results for subsequent searches synchronously', () => {
    boox.addDocumentsSync(datasets)

    const results1 = boox.searchSync('lorem ipsum')
    const results2 = boox.searchSync('lorem ipsum')

    expect(results1.length).toBe(1)
    expect(results2.length).toBe(1)
    expect(results1).toEqual(results2)
  })

  it('should retrieve search suggestions', async () => {
    await boox.addDocuments(datasets)

    const queries = ['document 1', 'document 2', 'document 3']
    queries.forEach(q => boox.searchSync(q))

    const suggestions = boox.getSearchSuggestions('doc')

    expect(suggestions).toEqual(['document 1', 'document 2', 'document 3'])
  })

  it('should retrieve search suggestions with threshold', async () => {
    await boox.addDocuments(datasets)

    const queries = ['document 1', 'document 3', 'document 2', 'document 3']
    queries.forEach(q => boox.searchSync(q))

    const suggestions = boox.getSearchSuggestions('doc', { threshold: 2 })

    expect(suggestions).toEqual(['document 3'])
  })

  it('should retrieve search suggestions with custom filter', async () => {
    await boox.addDocuments(datasets)

    const queries = ['document 1', 'document 3', 'document 2', 'document 3']
    queries.forEach(q => boox.searchSync(q))

    const suggestions = boox.getSearchSuggestions('doc', {
      limit: 3,
      filter(suggestion) {
        return !suggestion.includes('2')
      }
    })

    expect(suggestions).toEqual(['document 3', 'document 1'])
  })

  it('should retrieve the search history', async () => {
    await boox.search('lorem ipsum')
    boox.searchSync('dolor sit')

    const searchHistory = boox.getSearchHistory()

    expect(searchHistory).toEqual(['dolor sit', 'lorem ipsum'])
  })

  it('should retrieve the popular searches', async () => {
    await boox.search('lorem ipsum')
    boox.searchSync('lorem ipsum')
    boox.searchSync('dolor sit')

    const popularSearches = boox.getPopularSearches()

    expect(popularSearches).toEqual([
      ['lorem ipsum', 2],
      ['dolor sit', 1]
    ])
  })

  it('should paginate search results', () => {
    const results: SearchResult<{ title: string; content: string }>[] =
      Array.from({ length: 20 }, (_, i) => {
        return new SearchResult({
          query: 'Document',
          encoder: vi.fn() as never,
          document: {
            id: String(i) as never,
            attributes: {
              title: `Document ${i}`,
              content: 'Lorem ipsum dolor sit amet.'
            },
            magnitude: 1
          }
        })
      })

    const paginatedResults = Boox.paginateSearchResults(results, 2, 5)

    expect(paginatedResults.results.length).toBe(5)
    expect(paginatedResults.totalResults).toBe(20)
    expect(paginatedResults.totalPages).toBe(4)
    expect(paginatedResults.currentPage).toBe(2)
  })

  // Error handling

  it('should log an error when trying to add duplicate documents', async () => {
    await boox.addDocument(datasets[0])
    await boox.addDocument(datasets[0])

    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while adding document:\n',
      expect.any(Error)
    )
  })

  it('should log an error when trying to add duplicate documents synchronously', () => {
    boox.addDocumentSync(datasets[0])
    boox.addDocumentSync(datasets[0])

    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while adding document:\n',
      expect.any(Error)
    )
  })

  it('should log an error when trying to add a document with missing attributes', async () => {
    await boox.addDocument({} as never)
    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while adding document:\n',
      expect.any(Error)
    )
    expect(boox.currentState).toEqual({
      configs: {
        id: 'id',
        features: ['title', 'content'],
        attributes: ['url']
      },
      documents: {},
      features: {},
      popularSearches: {}
    })
  })

  it('should log an error when trying to add a document with missing attributes synchronously', () => {
    boox.addDocumentSync({} as never)
    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while adding document:\n',
      expect.any(Error)
    )
    expect(boox.currentState).toEqual({
      configs: {
        id: 'id',
        features: ['title', 'content'],
        attributes: ['url']
      },
      documents: {},
      features: {},
      popularSearches: {}
    })
  })

  it('should log an error when trying to update a document with missing attributes', async () => {
    await boox.updateDocument({} as never)

    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while updating document:\n',
      expect.any(Error)
    )
    expect(boox.currentState).toEqual({
      configs: {
        id: 'id',
        features: ['title', 'content'],
        attributes: ['url']
      },
      documents: {},
      features: {},
      popularSearches: {}
    })
  })

  it('should log an error when trying to update a document with missing attributes synchronously', () => {
    boox.updateDocumentSync({} as never)

    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while updating document:\n',
      expect.any(Error)
    )
    expect(boox.currentState).toEqual({
      configs: {
        id: 'id',
        features: ['title', 'content'],
        attributes: ['url']
      },
      documents: {},
      features: {},
      popularSearches: {}
    })
  })

  it('should log an error when search query expansion function throws an error', async () => {
    await boox.addDocument(datasets[0])

    const queryExpander = () => {
      throw new Error('Query expansion failed')
    }
    const results = await boox.search('lorem', { queryExpander })

    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while searching:\n',
      expect.any(Error)
    )
    expect(results).toHaveLength(0)
  })

  it('should log an error when search query expansion function throws an error synchronously', () => {
    boox.addDocumentSync(datasets[0])

    const queryExpander = () => {
      throw new Error('Query expansion failed')
    }
    const results = boox.searchSync('lorem', { queryExpander })

    expect(errorSpy).toHaveBeenCalledWith(
      'Error occurred while searching:\n',
      expect.any(Error)
    )
    expect(results).toHaveLength(0)
  })
})
