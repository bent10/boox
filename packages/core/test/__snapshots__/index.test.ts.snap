// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Boox > should add a document to the index 1`] = `
{
  "attributes": {
    "content": "Lorem ipsum dolor sit amet.",
    "title": "Document 1",
    "url": "/foo",
  },
  "id": "1",
  "magnitude": 0.8366600265340757,
}
`;

exports[`Boox > should add a document to the index synchronously 1`] = `
{
  "attributes": {
    "content": "Lorem ipsum dolor sit amet.",
    "title": "Document 1",
    "url": "/foo",
  },
  "id": "1",
  "magnitude": 0.8366600265340757,
}
`;

exports[`Boox > should add multiple documents to the index 1`] = `
{
  "attributes": {
    "content": "Lorem ipsum dolor sit amet.",
    "title": "Document 1",
    "url": "/foo",
  },
  "id": "1",
  "magnitude": 0.8366600265340757,
}
`;

exports[`Boox > should add multiple documents to the index 2`] = `
{
  "attributes": {
    "content": "Consectetur adipiscing elit.",
    "title": "Document 2",
    "url": "/bar",
  },
  "id": "2",
  "magnitude": 0.9128709291752769,
}
`;

exports[`Boox > should add multiple documents to the index 3`] = `
{
  "attributes": {
    "content": "Sed do eiusmod tempor incididunt.",
    "title": "Document 3",
    "url": "/baz",
  },
  "id": "3",
  "magnitude": 0.8366600265340757,
}
`;

exports[`Boox > should add multiple documents to the index synchronously 1`] = `
{
  "attributes": {
    "content": "Lorem ipsum dolor sit amet.",
    "title": "Document 1",
    "url": "/foo",
  },
  "id": "1",
  "magnitude": 0.8366600265340757,
}
`;

exports[`Boox > should add multiple documents to the index synchronously 2`] = `
{
  "attributes": {
    "content": "Consectetur adipiscing elit.",
    "title": "Document 2",
    "url": "/bar",
  },
  "id": "2",
  "magnitude": 0.9128709291752769,
}
`;

exports[`Boox > should add multiple documents to the index synchronously 3`] = `
{
  "attributes": {
    "content": "Sed do eiusmod tempor incididunt.",
    "title": "Document 3",
    "url": "/baz",
  },
  "id": "3",
  "magnitude": 0.8366600265340757,
}
`;

exports[`Boox > should handle searching with advanced similarity scoring 1`] = `
[
  SearchResult {
    "attributes": {
      "content": "Lorem ipsum dolor sit amet.",
      "title": "Document 1",
      "url": "/foo",
    },
    "highlightTag": [
      "<mark class="search-highlight">",
      "</mark>",
    ],
    "id": "1",
    "magnitude": 0.8366600265340757,
    "query": "document",
    "score": 0.5342384456611643,
    "vector": null,
  },
  SearchResult {
    "attributes": {
      "content": "Consectetur adipiscing elit.",
      "title": "Document 2",
      "url": "/bar",
    },
    "highlightTag": [
      "<mark class="search-highlight">",
      "</mark>",
    ],
    "id": "2",
    "magnitude": 0.9128709291752769,
    "query": "document",
    "score": 0.5342384456611643,
    "vector": null,
  },
  SearchResult {
    "attributes": {
      "content": "Sed do eiusmod tempor incididunt.",
      "title": "Document 3",
      "url": "/baz",
    },
    "highlightTag": [
      "<mark class="search-highlight">",
      "</mark>",
    ],
    "id": "3",
    "magnitude": 0.8366600265340757,
    "query": "document",
    "score": 0.5342384456611643,
    "vector": null,
  },
]
`;

exports[`Boox > should handle searching with advanced similarity scoring synchronously 1`] = `
[
  SearchResult {
    "attributes": {
      "content": "Lorem ipsum dolor sit amet.",
      "title": "Document 1",
      "url": "/foo",
    },
    "highlightTag": [
      "<mark class="search-highlight">",
      "</mark>",
    ],
    "id": "1",
    "magnitude": 0.8366600265340757,
    "query": "document",
    "score": 0.5342384456611643,
    "vector": null,
  },
  SearchResult {
    "attributes": {
      "content": "Consectetur adipiscing elit.",
      "title": "Document 2",
      "url": "/bar",
    },
    "highlightTag": [
      "<mark class="search-highlight">",
      "</mark>",
    ],
    "id": "2",
    "magnitude": 0.9128709291752769,
    "query": "document",
    "score": 0.5342384456611643,
    "vector": null,
  },
  SearchResult {
    "attributes": {
      "content": "Sed do eiusmod tempor incididunt.",
      "title": "Document 3",
      "url": "/baz",
    },
    "highlightTag": [
      "<mark class="search-highlight">",
      "</mark>",
    ],
    "id": "3",
    "magnitude": 0.8366600265340757,
    "query": "document",
    "score": 0.5342384456611643,
    "vector": null,
  },
]
`;

exports[`Boox > should retrieve the current state of the Boox instance 1`] = `
{
  "configs": {
    "attributes": [
      "url",
    ],
    "features": [
      "title",
      "content",
    ],
    "id": "id",
  },
  "documents": {
    "1": {
      "attributes": {
        "content": "Lorem ipsum dolor sit amet.",
        "title": "Document 1",
        "url": "/foo",
      },
      "id": "1",
      "magnitude": 0.8366600265340757,
    },
    "2": {
      "attributes": {
        "content": "Consectetur adipiscing elit.",
        "title": "Document 2",
        "url": "/bar",
      },
      "id": "2",
      "magnitude": 0.9128709291752769,
    },
    "3": {
      "attributes": {
        "content": "Sed do eiusmod tempor incididunt.",
        "title": "Document 3",
        "url": "/baz",
      },
      "id": "3",
      "magnitude": 0.8366600265340757,
    },
  },
  "features": {
    "content": {
      "adipiscing": {
        "2": 0.3333333333333333,
      },
      "amet": {
        "1": 0.2,
      },
      "consectetur": {
        "2": 0.3333333333333333,
      },
      "do": {
        "3": 0.2,
      },
      "dolor": {
        "1": 0.2,
      },
      "eiusmod": {
        "3": 0.2,
      },
      "elit": {
        "2": 0.3333333333333333,
      },
      "incididunt": {
        "3": 0.2,
      },
      "ipsum": {
        "1": 0.2,
      },
      "lorem": {
        "1": 0.2,
      },
      "sed": {
        "3": 0.2,
      },
      "sit": {
        "1": 0.2,
      },
      "tempor": {
        "3": 0.2,
      },
    },
    "title": {
      "1": {
        "1": 0.5,
      },
      "2": {
        "2": 0.5,
      },
      "3": {
        "3": 0.5,
      },
      "document": {
        "1": 0.5,
        "2": 0.5,
        "3": 0.5,
      },
    },
  },
  "popularSearches": {},
}
`;

exports[`Boox > should update an existing document in the index 1`] = `
{
  "attributes": {
    "content": "Updated content.",
    "title": "Updated Document Title",
    "url": "/foo",
  },
  "id": "1",
  "magnitude": 0.9128709291752768,
}
`;

exports[`Boox > should update an existing document in the index synchronously 1`] = `
{
  "attributes": {
    "content": "Updated content.",
    "title": "Updated Document Title",
    "url": "/foo",
  },
  "id": "1",
  "magnitude": 0.9128709291752768,
}
`;
